crates/ra_ide_api/src/name_kind.rs

```diff
@@ -0,0 +1,316 @@
+//! FIXME: write short doc here
+
+use hir::{
+    db::AstDatabase, Adt, AssocItem, DefWithBody, Either, EnumVariant, FromSource, HasSource,
+    HirFileId, MacroDef, ModuleDef, ModuleSource, Path, PathResolution, SourceAnalyzer,
+    StructField, Ty, VariantDef,
+};
+use ra_db::FileId;
+use ra_syntax::{ast, ast::VisibilityOwner, AstNode, AstPtr};
+
+use crate::db::RootDatabase;
+
+pub enum NameKind {
+    Macro(MacroDef),
+    FieldAccess(StructField),
+    AssocItem(AssocItem),
+    Def(ModuleDef),
+    SelfType(Ty),
+    Pat((DefWithBody, AstPtr<ast::BindPat>)),
+    SelfParam(AstPtr<ast::SelfParam>),
+    GenericParam(u32),
+}
+
+pub(crate) struct Declaration {
+    visibility: Option<ast::Visibility>,
+    container: ModuleSource,
+    pub item: NameKind,
+}
+
+trait HasDeclaration {
+    type Def;
+    type Ref;
+
+    fn declaration(self, db: &RootDatabase) -> Declaration;
+    fn from_def(db: &RootDatabase, file_id: HirFileId, def: Self::Def) -> Option<Declaration>;
+    fn from_ref(
+        db: &RootDatabase,
+        analyzer: &SourceAnalyzer,
+        refer: Self::Ref,
+    ) -> Option<Declaration>;
+}
+
+macro_rules! match_ast {
+    (match $node:ident {
+        $( ast::$ast:ident($it:ident) => $res:block, )*
+        _ => $catch_all:expr,
+    }) => {{
+        $( if let Some($it) = ast::$ast::cast($node.clone()) $res else )*
+        { $catch_all }
+    }};
+}
```

Comment:
Sweet! Could you add this to `ra_syntax`, and remove visitors infra in a separate PR? this will close #1672

Reply:
Oh, and I think it's more flexible to use `:expr` instead of `:ident` for node, but care must be taken to evaluate the expression only once

Reply:
I think :expr is not allowd because of the following {.

Reply:
Yeah, that's right: `{` can't follow `:expr`.
There's already an open PR for the issue, that's why I didn't add it separately.

Reply:
right, it's fine to stick with `$ident` then I think. That PR seems stalled, so it's fine to open a new one